x/batching/keeper/endblock.go

```diff
@@ -49,21 +52,44 @@ func (k Keeper) EndBlock(ctx sdk.Context) (err error) {
 // It returns a resulting batch, data result tree entries, and validator
 // tree entries in that order.
 func (k Keeper) ConstructBatch(ctx sdk.Context) (types.Batch, [][]byte, [][]byte, error) {
-	// Note current will be "old" for this new batch.
-	oldBatchNum, err := k.GetCurrentBatchNum(ctx)
+	var newBatchNum uint64
+	var latestDataRoot, latestValRoot string
+	latestBatch, err := k.GetLatestBatch(ctx)
 	if err != nil {
-		return types.Batch{}, nil, nil, err
+		if !errors.Is(err, types.ErrBatchingHasNotStarted) {
+			return types.Batch{}, nil, nil, err
+		}
+		newBatchNum = collections.DefaultSequenceStart + 1
+	} else {
+		newBatchNum = latestBatch.BatchNumber + 1
+		latestDataRoot = latestBatch.DataResultRoot
+		latestValRoot = latestBatch.ValidatorRoot
 	}
-	newBatchNum := oldBatchNum + 1
 
+	// Compute current data result tree root and the "super root"
+	// of current and previous data result trees' roots.
 	dataEntries, dataRoot, err := k.ConstructDataResultTree(ctx, newBatchNum)
 	if err != nil {
 		return types.Batch{}, nil, nil, err
 	}
+	latestDataRootHex, err := hex.DecodeString(latestDataRoot)
+	if err != nil {
+		return types.Batch{}, nil, nil, err
+	}
+	superRoot := utils.RootFromLeaves([][]byte{latestDataRootHex, dataRoot})
+
+	// Compute validator tree root.
 	valEntries, valRoot, err := k.ConstructValidatorTree(ctx)
 	if err != nil {
 		return types.Batch{}, nil, nil, err
 	}
+	valRootHex := hex.EncodeToString(valRoot)
+
+	// Skip batching if there is no update in data result root nor
+	// validator root.
+	if len(dataEntries) == 0 && valRootHex == latestValRoot {
+		return types.Batch{}, nil, nil, types.ErrNoBatchingUpdate
+	}
```

Comment:
Not sure if it's worth it but we could move the DR root construction after this check.

Reply:
I don't think that's worth it because if there are no data results to be batched, the method `ConstructDataResultTree` would be very light anyways.